fix(automation): fix cyclic dependencies and parallel build in topological order

rarajes2 opened the pull request:
<!--
Hey there,\
Thank you for taking the time to improve our code! 🙂\
Please let us know why this change is necessary and what testing you have performed. \
This ensures our reviewers understand the impact of your change. \

**IMPORTANT**
FAILING TO FILL OUT THIS TEMPLATE WILL RESULT IN REJECTION OF YOUR PULL REQUEST
This is for compliance purposes with FedRAMP program.
-->

# COMPLETES #< pre - [CAI-6612](https://jira-eng-sjc12.cisco.com/jira/browse/CAI-6612) >

## This pull request addresses

Fixes the cyclic dependencies and updates the PR workflow to build the packages parallelly in topological order.

## by making the following changes

Changing the name moves the package up in the list and when ran in sequence it fails due to dependency on other package which are yet to be built.

Adding `--topological` ensures the dependencies are built before building the package.

<!-- You may include screenshots -->

### Change Type

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [x] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

## The GAI Coding Policy And Copyright Annotation Best Practices ##
 
<!-- **MANDATORY** If Yes, Mention the GAI Coding Policy Copyright Annotation Best Practices followed separated by a comma below the yes checkbox -->
 
- [ ] GAI was not used (or, no additional notation is required)
- [ ] Code was generated entirely by GAI
- [ ] GAI was used to create a draft that was subsequently customized or modified
- [ ] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [x] Tool used for AI assistance (GitHub Copilot / Other - specify)
  - [ ] Github Copilot
  - [x] Other - Windsurf
- [ ] This PR is related to
  - [ ] Feature
  - [ ] Defect fix
  - [ ] Tech Debt
  - [ ] Automation

### I certified that

- [ ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request
- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

rarajes2 committed:
fix(automation): fix the other packages build

rarajes2 committed:
Merge branch 'next' into update-workflow

rarajes2 committed:
fix(automation): fix cyclic dependencies

rarajes2 committed:
fix(automation): fix integration tests dependencies

rarajes2 commented:
Integration test is failing in the next branch too.

Kesari3008 commented on packages/@webex/plugin-attachment-actions/package.json:
```diff
@@ -48,6 +48,9 @@
     "@webex/eslint-config-legacy": "workspace:*",
     "@webex/jest-config-legacy": "workspace:*",
     "@webex/legacy-tools": "workspace:*",
+    "@webex/plugin-logger": "workspace:*",
```
Why did we need to add these dependencies as peer dependency as well as dev dependency ? Also why do we not need @webex/test-helper-chai, @webex/test-helper-test-users.

rarajes2 replied:
Dependencies added as peer are required when added to application (avoiding copies), the same dependencies are added a dev dependencies due to being used in the integration tests. Please check the code rabbit's comment.

test dependencies are removed as they were duplicate. They are already present in the dev dependencies.

Kesari3008 commented on packages/@webex/plugin-memberships/package.json:
```diff
@@ -36,14 +36,16 @@
     "debug": "^4.3.4",
     "lodash": "^4.17.21"
   },
+  "peerDependencies": {
```
Was this change needed ?

rarajes2 replied:
Have addressed [here](https://github.com/webex/webex-js-sdk/pull/4428#discussion_r2272011016)

Kesari3008 reviewed (commented):
I see some of the packages being moved from devDependecies to peer dependencies and some of the plugins are just removed. Do we  not need the removed dependencies anymore ?

Please update the description as it says that it has renaming changing but it doesn't. Maybe we can elaborate on that in order to do so, cyclic dependencies needed to be fixed and that's what this PR addresses

rarajes2 commented:
> I see some of the packages being moved from devDependecies to peer dependencies and some of the plugins are just removed. Do we not need the removed dependencies anymore ?
> 
> Please update the description as it says that it has renaming changing but it doesn't. Maybe we can elaborate on that in order to do so, cyclic dependencies needed to be fixed and that's what this PR addresses

Removed dependencies were not being imported anywhere in the package. Updated the description.

rarajes2 committed:
fix(automation): add broken to integration test for webex package

rarajes2 committed:
Merge branch 'next' into update-workflow

rarajes2 committed:
fix(automation): add dev dependencies for webex-core package

Shreyas281299 commented on .github/workflows/pull-request.yml:
```diff
@@ -160,7 +160,7 @@ jobs:
 
       - name: Build Other Packages
-        run: yarn workspaces foreach --parallel --verbose run build:src
+        run: yarn workspaces foreach --parallel --topological --verbose run build:src
```
How have we tested this? This might break the pipeline if not tested properly (You can run this in your fork and then share a github action successfull run). Also what about deploy.yml? Wouldnt that also need changes? And the main package.json we are doing build:local and prebuild:module shouldnt we be making this change there also?

rarajes2 replied:
Tested this on the fork repo. PR - https://github.com/rarajes2/webex-js-sdk/pull/3

Updated the `deploy.yml` and commands in `package.json` file such as `build:local`, `build:publish`. 

Let's discuss `prebuild:modules` command.

rarajes2 replied:
As discussed, updated and tested the `prebuild:modules` command as well.

Shreyas281299 replied:
Looks good

rarajes2 committed:
fix(automation): update package.json commands and deploy.yml

jor-row reviewed (approved):
approved but tests are not passing

rarajes2 commented:
> approved but tests are not passing

@jor-row - Those tests are failing in the next branch as well.

rarajes2 committed:
Merge branch 'next' into update-workflow
